### PR TITLE
Handle update of an exported service

### LINKED ISSUE #645: Updating an exported service is not handled
**What happened**:
In a working setup, when a service is exported, we can successfully access it from remote clusters.
If we update the service (like the port numbers) in the origin cluster, the corresponding serviceImport is not updated.

### PR DESCRIPTION
**Pull request by blue-troy**
When an exported service update, Service deletion
will trigger ServiceExport update.

Fixes #645
Signed-off-by: blue-troy <12729455+blue-troy@users.noreply.github.com>

### CODE REVIEW

File: pkg/agent/controller/agent.go
Comment (tpantelis) [inline pkg/agent/controller/agent.go]: What specific modification of an exported service would warrant this action? It seems potentially disruptive to delete the `ServiceImport` - why not update the existing one?

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/agent.go
Comment (blue-troy) [reply pkg/agent/controller/agent.go]: The service may change to an unsupported type service. service Import seem don't have a not avalible status

Comment (tpantelis) [reply pkg/agent/controller/agent.go]: Then let's handle that specific case rather than blindly deleting the `ServiceImport`.

Comment (blue-troy) [reply pkg/agent/controller/agent.go]: Good idea, i will try to solve it in the next few days

Comment (blue-troy) [reply pkg/agent/controller/agent.go]: update `ServiceExport` status in Service deletion, let `ServiceExport -> ServiceImport` decide how to deal with it.

File: pkg/agent/controller/agent.go
Comment (tpantelis) [inline pkg/agent/controller/agent.go]: Rather than changing the status condition to an intermediate state indicating it's not valid and relying on a `ServiceExport` update to trigger `serviceExportToServiceImport`, I think it would be cleaner to call `serviceExportToServiceImport`  directly:

`return serviceExportToServiceImport(svcExport, numRequeues, syncer.Update)`

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/agent.go
Comment (blue-troy) [reply pkg/agent/controller/agent.go]: When call `return serviceExportToServiceImport(svcExport, numRequeues, syncer.Update)` directly

[serviceExportToServiceImport code](https://github.com/submariner-io/lighthouse/blob/v0.14.0-m1/pkg/agent/controller/agent.go#L263)

```go
	if op == syncer.Update && getLastExportConditionReason(svcExport) != serviceUnavailable {
		return nil, false
	}
```

LastExportConditionReason "" may be trigger by `onSuccessfulServiceImportSync` or `serviceToRemoteServiceImport` directly call at least. will ConditionReason "" may be somthing else? If it's not,we can change onSuccessfulServiceImportSync reason with `ServiceImportSynced`.

If we do not want to change `onSuccessfulServiceImportSync` reason, change `ServiceExport` then directly call `serviceToRemoteServiceImport` will cause `serviceToRemoteServiceImport` triggered more than it's needs.

Comment (tpantelis) [reply pkg/agent/controller/agent.go]: Sorry I'm not sure I follow. So if `serviceToRemoteServiceImport` is called directly then the above check will skip updating the `ServiceImport`? If so then perhaps instead of checking the last condition reason we check if the last condition Status is `ConditionTrue`, ie process the update only if the service was previously exported.

Comment (blue-troy) [reply pkg/agent/controller/agent.go]: you mean [serviceExportToServiceImport](https://github.com/submariner-io/lighthouse/blob/v0.14.0-m1/pkg/agent/controller/agent.go#L263) like this ?：

```go
	if op == syncer.Update && getLastExportCondition(svcExport) != corev1.ConditionTrue {
		return nil, false
	}
```

onSuccessfulServiceImportSync will trigger this ConditionTrue too. how distinguish service update direct call and onSuccessfulServiceImportSync? service update direct call and onSuccessfulServiceImportSync has the same Status and Reason

Comment (tpantelis) [reply pkg/agent/controller/agent.go]: https://github.com/submariner-io/lighthouse/pull/899 should alleviate your issues with calling `serviceToRemoteServiceImport` directly.

Comment (blue-troy) [reply pkg/agent/controller/agent.go]: OK, i will have a look

Comment (blue-troy) [reply pkg/agent/controller/agent.go]: #899 is useful in our case 👍, i will push direct call `serviceToRemoteServiceImport` version code after your  #899 PR merged

**Comment by blue-troy** [comment]
should we update [code](https://github.com/submariner-io/lighthouse/blob/8a8c1fbfc8d34b98c15aaef670d1c3797cdb2fbf/pkg/agent/controller/agent.go#L269-L277) like this ?:

```go
	svcType, ok := getServiceImportType(svc)

	if !ok {
		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, mcsv1a1.ServiceExportValid, corev1.ConditionFalse,
			invalidServiceType, fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
		klog.Errorf("Service type %q not supported", svc.Spec.Type)

		//delete unsupported service type
		err := a.serviceImportSyncer.GetLocalFederator().Delete(serviceImport)
                 ....
}
```

to support case like an exported service type change from a supported type to an unsupported type?

**Comment by vthapar** [comment]
> should we update [code](https://github.com/submariner-io/lighthouse/blob/8a8c1fbfc8d34b98c15aaef670d1c3797cdb2fbf/pkg/agent/controller/agent.go#L269-L277) like this ?:
> 
> ```go
> 	svcType, ok := getServiceImportType(svc)
> 
> 	if !ok {
> 		a.updateExportedServiceStatus(svcExport.Name, svcExport.Namespace, mcsv1a1.ServiceExportValid, corev1.ConditionFalse,
> 			invalidServiceType, fmt.Sprintf("Service of type %v not supported", svc.Spec.Type))
> 		klog.Errorf("Service type %q not supported", svc.Spec.Type)
> 
> 		//delete unsupported service type
> 		err := a.serviceImportSyncer.GetLocalFederator().Delete(serviceImport)
>                  ....
> }
> ```
> 
> to support case like an exported service type change from a supported type to an unsupported type?

It should be done becasue a serviceImport shouldn't exist for a service that isn't supported as per current logic. But would like to add a check for `op.Update`

@tpantelis In case of delete of something that doesn't exist yet, do we silently ignore, or log it?

File: pkg/agent/controller/agent.go
Comment (vthapar) [inline pkg/agent/controller/agent.go]: Why changing this to false?

**Review by vthapar — CHANGES_REQUESTED** [review]

File: pkg/agent/controller/agent.go
Comment (blue-troy) [reply pkg/agent/controller/agent.go]: serviceToRemoteServiceImport can cover `service change from not supported type to supported type` case, by call a.serviceExportToServiceImport.

Comment (tpantelis) [reply pkg/agent/controller/agent.go]: It should. We should add unit test cases to cover the new changes. I can do that in a follow-up PR.

**Comment by tpantelis** [comment]
> It should be done becasue a serviceImport shouldn't exist for a service that isn't supported as per current logic. But would like to add a check for `op.Update`
> 
> @tpantelis In case of delete of something that doesn't exist yet, do we silently ignore, or log it?

It doesn't: https://github.com/submariner-io/admiral/blob/e07dc24fda439c217cf8c512161474b1038aa355/pkg/federate/base_federator.go#L70. It's up to the caller to check the error.

**Comment by tpantelis** [comment]
OK I pushed 2 additional commits: the first deletes the `ServiceImport` when the type is unsupported and the second adds unit tests cases to cover the changes. I think we should get this in for 0.14. @vthapar @aswinsuryan 

**Edit**: also pushed commit for the release note.

**Review by tpantelis — APPROVED** [review]

**Review by dfarrell07 — APPROVED** [review]
LGTM, thanks @blue-troy!

**Review by skitt — APPROVED** [review]

**Comment by skitt** [comment]
@vthapar this is waiting on your re-review.

**Review by vthapar — APPROVED** [review]
Thanks for all the work on this @blue-troy and @tpantelis . Looks good.

**Pull request merged**
